Framework: ES Modules + Tree Shaking

samouri opened the pull request:
Webpack2 and Webpack 3 introduced a couple major features:

1. [native es6 modules + tree shaking](https://medium.com/webpack/webpack-2-and-beyond-40520af9067f).  This promises to make our modules smaller, compile times faster (less babel transpilation), and our code more correct (greater compile time checks).
2. [Scope Hoisting](https://medium.com/webpack/webpack-3-official-release-15fd2dd8f07b).  This feature is about making the code smaller and the runtime properties faster by removing wepback wrapper funcs in favor of direct inlining when possible.

Currently, we cant take advantage of either because we are still [transpiling](https://github.com/Automattic/wp-calypso/blob/master/.babelrc) to commonjs modules in our babel configuration.

In this PR I try to  find (and hopefully solve) all of the blockers towards disabling commonjs transpilation and enabling tree-shaking/scope hoisting.


To Test
---
Test everything on: https://dserve.a8c.com/?branch=try/scope-hoisting.

samouri commented:
going to start this guy from scratch now!

samouri commented:
Turns out we are misusing `export`/`import` all over the place in calypso.

Heres an example error from webpack:
```
WARNING in ./client/boot/common.js
160:50-65 "export 'isDefaultLocale' was not found in 'lib/i18n-utils'
```

The reason that is an error is because in es6 you cannot destructure object imports.
This problem is relatively pervasive in our codebase.  Heres a bigger explanation:

**Destructuring Imports**
In proper es6 you _cannot_ do a destructing import of an object. That means if you write:
`import {thing1, thing2} from 'lib'` then `thing` and `thing2` must be actual named exports and not just keys in an export object.  

This would a syntactically correct lib file:
```js
// correct lib implementation
export const thing1 = function (){ ... };
export const thing2 = function (){ ... };
```

This is an incorrect one:
```js
// incorrect lib implementation
export default {
  thing1: function(){ ... },
  thing2: function(){ ... }
};
```


Here is a list i generated of offending files and the frequency with which their exports are being used incorrectly. Note all relative urls need more investigation.
```
  11 '../'
   2 '../comment-replies-cache'
   3 './assembler'
   1 './comment-replies-cache'
  66 './constants'
   5 './controller'
  10 './data'
   1 './helpers'
   3 './schema'
   1 './services'
  15 './utils'
   1 'components/phone-input/data'
   1 'components/post-schedule/utils'
  46 'config'
  31 'lib/analytics'
  10 'lib/analytics/ad-tracking'
 101 'lib/cart-values'
  12 'lib/cart-values/cart-items'
   1 'lib/cart/store/cart-analytics'
   3 'lib/countries-list'
   5 'lib/credit-card-details'
   2 'lib/data-poller'
   1 'lib/domains/assembler'
  58 'lib/domains/constants'
   3 'lib/domains/email-forwarding'
   2 'lib/domains/google-apps-users'
   4 'lib/domains/whois/utils'
   1 'lib/feature-detection'
   5 'lib/feed-post-store/actions'
   3 'lib/feed-post-store/constants'
   1 'lib/feed-post-store/display-types'
   3 'lib/feed-stream-store/constants'
 112 'lib/formatting'
  18 'lib/i18n-utils'
   2 'lib/impure-lodash'
   2 'lib/like-store/actions'
   1 'lib/media-serialization'
   1 'lib/media/store'
  10 'lib/media/utils'
  32 'lib/paths'
   1 'lib/plugins/utils'
  12 'lib/posts/utils'
 208 'lib/products-values'
  17 'lib/route'
  23 'lib/route/path'
   3 'lib/shortcode'
   1 'lib/signup/actions'
   2 'lib/signup/step-actions'
  18 'lib/site/utils'
   4 'lib/sites-list/actions'
   5 'lib/store-transactions'
   3 'lib/touch-detect'
  38 'lib/upgrades/actions'
  73 'lib/url'
   4 'lib/url/support'
   1 'lib/users/actions'
  24 'lib/viewport'
   6 'me/event-recorder'
   4 'me/purchases/paths'
  54 'my-sites/controller'
  22 'my-sites/domains/paths'
   4 'my-sites/invites/utils'
   2 'signup/config/flows'
   1 'signup/utils'
   3 'state/action-types'
   8 'state/current-user/constants'
  19 'state/jetpack-connect/actions'
  20 'state/jetpack-connect/selectors'
   2 'state/jetpack-sync/actions'
   3 'state/plugins/premium/actions'
  18 'state/plugins/premium/selectors'
   5 'state/plugins/wporg/actions'
   7 'state/plugins/wporg/selectors'
  55 'woocommerce/lib/nav-utils'
```

samouri commented:
[bad-imports.txt](https://github.com/Automattic/wp-calypso/files/1367294/bad-imports.txt)

samouri commented:
next two things:
1. try out named-exports codemod
2. fix leftovers by hand
3. profit?

samouri commented:
found an issue in notifications-panel.  I create a PR there and until its merged will need to test with: 
`npm install Automattic/notifications-panel#fix/imports-for-tree-shaking`

samouri commented:
Really cool news about the stricter es6 syntax: the code will be able to verify a greater amount of correctness _at compile time_.

Now the interesting bit, I found 8 issues that are most likely bugs in production right now:

1. [ ] WARNING in ./client/state/notices/middleware.js
    214:1724-1752 "export 'PUBLICIZE_CONNECTION_REFRESH' was not found in 'state/action-types'
2. [ ] WARNING in ./client/state/login/magic-login/reducer.js
     23:8-25 "export 'INTERSTITIAL_PAGE' was not found in './constants'
3. [ ] WARNING in ./client/state/signup/optional-dependencies/reducer.js
     14:4-27 "export 'suggestedUsernameSchema' was not found in './schema'
4. [ ] WARNING in ./client/state/site-settings/exporter/reducers.js
    97:7-36 "export 'EXPORT_ADVANCED_SETTINGS_FAIL' was not found in 'state/action-types'
5. [ ] WARNING in ./client/state/sharing/publicize/publicize-actions/reducer.js
    61:21-43 "export 'publicizeActionsSchema' was not found in './schema'
6. [ ] WARNING in ./client/state/site-settings/exporter/reducers.js
    97:7-36 "export 'EXPORT_ADVANCED_SETTINGS_FAIL' was not found in 'state/action-types'
7. [ ] WARNING in ./client/reader/tag-stream/index.js
    17:55-70 "export 'recommendedTags' was not found in './controller'
8. [ ] WARNING in ./client/my-sites/sharing/connections/services-group.jsx
    44:20-45 "export 'hasOwnProperty' (imported as 'Components') was not found in './services'
    NOTE: this last one isn't a bug. it just won't be legal in es6land
```

samouri commented:
A few more notes from me whacking at the issue today:

1. we need to get rid of all requires and use imports instead. that has ramifications in a few areas:
1a. both sections loader and extensions-loader (reducers) need changes
1.2 anywhere we are relying upon dead-code elimination and nested requires we will need a different strategy.  one idea could be: `export default ( env === 'production' ? noop : actualexport` at relevant location instead of at the import site but its less natural and harder to track down which are dependent on env.  I'll do research to see if theres a better way.
1.3  also theres no codemod for this, so its a lot of hand-edits 😢 .
2. - [ ] change required in [notifications-panel](https://github.com/Automattic/notifications-panel/pull/184)

DanReyLop commented:
> 1.3 anywhere we are relying upon dead-code elimination and nested requires we will need a different strategy. one idea could be: export default ( env === 'production' ? noop : actualexport at relevant location instead of at the import site but its less natural and harder to track down which are dependent on env. I'll do research to see if theres a better way.

Relevant: https://github.com/webpack/webpack/issues/2061

If I understood it correctly, Webpack does a basic pass to check if an `import` is really used, and then UglifyJS does the actual code elimination. So, here:

```js
import { otherFunc1, otherFunc2 } from './otherModule';

function callOther() {
  otherFunc1();
  otherFunc2();
}

export default function () {
  if (always_false) {
    callOther();
  }
}
```

Webpack will assume that `otherFunc1` and `otherFunc2` are being used. UglifyJS then will remove the `if` and the `callOther` function entirely, but will keep the `import`.

But in this case:
```js
import { otherFunc1, otherFunc2 } from './otherModule';

export default function () {
  if (always_false) {
    otherFunc1();
    otherFunc2();
  }
}
```

Webpack correctly detects that `otherFunc1` and `otherFunc2` aren't being used, so trees will be shaken and the `import` will be removed entirely.

I think our use cases can be adapted to the second case, just use the `import`ed module inside a constant conditional and it will be optimized away.

For the first case, I assume that if the `import`ed module is part of the same bundle, and scope hoisting is fully functioning, then everything can be optimized away by UglifyJS too, but I have nothing solid to back this up.

psealock commented:
@samouri I pushed a commit directly to this branch 7da6bc5bd59f2d4bd338f97d0a62f3be1fb63ab0

>Next step is to take all the meaty import/export commits and cherry-pick them into a different PR

This one's not too meaty, but I can do this as well.

Tests were failing due this type of code
```js
import initialState from 'state/initial-state';
...
const createReduxStoreFromPersistedInitialState = initialState.default;
```
### Fix
```js
import theDefault, { otherFunc1, otherFunc2 } from 'state/initial-state';
```
confirmed the non-test import is correctly written

gziolo commented on client/boot/app.js:
```diff
@@ -20,13 +17,15 @@ import { configureReduxStore, locales, setupMiddlewares, utils } from './common'
 	debug( "Starting Calypso. Let's do this." );
 
-	const project = require( `./project/${ PROJECT_NAME }` );
+	const project = projects[ PROJECT_NAME ];
```
It is a different implementation. It will always load the project which is s valid use case for WordPress.com, but it prevents us from making it customizable. The idea here was to load only code for a one project specified in config file. Webpack optimizes everything behind the scenes.

gziolo commented on client/boot/app.js:
```diff
@@ -1,10 +1,7 @@
-if ( process.env.NODE_ENV === 'development' ) {
-	require( 'lib/wrap-es6-functions' )();
-}
+import 'lib/wrap-es6-functions';
```
It would be nice to remove it completely, which should be possible since we started using babel preset env.

DanReyLop replied:
We aren't using `babel-preset-env` to polyfill yet, though. We're waiting for `babel-preset-env@2`, which can include only the polyfills that actually used & required for the codebase.

gziolo replied:
So it’s fine as it is 👍

gziolo commented on client/boot/common.js:
```diff
@@ -76,7 +81,7 @@ const setupContextMiddleware = reduxStore => {
 };
 
 // We need to require sections to load React with i18n mixin
```
We can remove comment if the mixin was removed 😃

gziolo commented on client/boot/dev-modules.js:
```diff
@@ -2,13 +2,16 @@
 	}
-}
+};
+export default ( process.env.NODE_ENV === 'development' ? cssReload : noop );
```
Nice 👍

gziolo commented on client/boot/common.js:
```diff
@@ -114,7 +118,7 @@ const loggedOutMiddleware = currentUser => {
 	if ( config.isEnabled( 'oauth' ) ) {
 		// Forces OAuth users to the /login page if no token is present
-		page( '*', require( 'auth/controller' ).checkToken );
+		page( '*', oathController.checkToken );
```
It’s interesting use case. It is bundled despite being behind the feature flag at the moment so it can be imported at the top ofvthe file

DanReyLop replied:
Idea for the future: Create a small Webpack plugin (similar to `DefinePlugin`) that resolves `config.blah()` calls at compile time. Could save more than a few KB :)

gziolo replied:
I had exactly the same idea at the first glance :) However it’s more complex as it seems. We keep it in this shape on purpose, because it allows us to create one build file that works with all non-development envs. We expose config data from the sever using JS variable instantiated inside the script tag. It was a trade-off that allowed us to speed up docker build and thus deploy time by a few minutes. Jake knows all the rationale behind it much better 😃

gziolo replied:
To be clear. I’m not saying we shouldn’t investigate it. It can be quite cool. We just need to evaluate if we gain enough that would justify having a dedicated build for Calypso env equal production.

DanReyLop replied:
Oh, that's right, I remember now that there were different builds before. Nevermind then, it would be a step backwards :)

Altough I agree that a separate `production` build would be worth it if we can make as many time-consuming tricks to optimize for bundle size as possible.

samouri replied:
We gain a lot by having all of our environments be based off the same docker image.  That said, I'm open to breaking with that if the tradeoffs are worth it!  They weren't in the past.

For this case, where we have code that is used in stages that aren't production, I think there is a more simple solution involving separating out that code into its own chunk. We do this right now with the devdocs chunk which is only actually used in the wpcalypso stage even though its accessible in all of them.

gziolo commented on client/lib/wrap-es6-functions/index.js:
```diff
@@ -28,12 +28,18 @@ function wrapObjectFn( obj, objectName, key ) {
 
-export default function() {
+const wrapEs6Functions = function() {
+	if ( process.env.NODE_ENV === 'development' ) {
```
If this code stats, it will do nothing. Both env guards are mutually exclusive 😃

samouri commented:
@gziolo: up until now this pr has been mostly exploratory.  I don't plan on merging the majority of the code written here.  I have been splitting out most of the key steps into separate PRs in order to keep the scope somewhat manageable and the code reviewable. 

For example: the 7 PRs referenced [here](https://github.com/Automattic/wp-calypso/pull/16057#issuecomment-335757214)

The current piece I am working on that needs review is to fix all the build errors: https://github.com/Automattic/wp-calypso/pull/18838

samouri commented:
https://dserve.a8c.com/?branch=try/scope-hoisting

samouri commented:
wow
---
```
Commit: af5e419aaa67114b5a48901c1f7fbeb8fc4ea6f0
Author: samouri
At: 2018-01-17T21:01:47Z
Message: hackety hack the extensions reducer

Delta:
chunk                                                       stat_size            parsed_size            gzip_size
accept-invite                                                -34834 B  (-16.8%)      -9062 B   (-7.8%)     -627 B   (-2.6%)
account-recovery                                             -22300 B  (-27.1%)      -6853 B  (-14.9%)    -1071 B  (-12.2%)
ads                                                          -26560 B  (-19.0%)      -7727 B   (-9.7%)     -245 B   (-1.6%)
async-load-blocks-calendar-popover                           -11285 B   (-3.3%)      -2712 B   (-1.6%)      -61 B   (-0.2%)
async-load-blocks-reader-full-post                           -22600 B  (-19.9%)      -6326 B  (-10.2%)     -423 B   (-3.1%)
async-load-components-happychat                               -9270 B  (-12.2%)      -3872 B  (-11.2%)     -382 B   (-4.3%)
async-load-components-web-preview-component                  -98381 B  (-18.9%)     -25535 B   (-9.4%)    -1567 B   (-2.8%)
async-load-components-webpack-build-monitor                    -966 B  (-17.1%)       -117 B   (-4.0%)       -3 B   (-0.2%)
async-load-extensions-woocommerce-app-store-stats            -25994 B   (-8.0%)     -31215 B  (-14.9%)    -7146 B  (-14.9%)
async-load-extensions-woocommerce-app-store-stats-listview   -11290 B  (-25.1%)      -2517 B  (-10.2%)      -72 B   (-1.4%)
async-load-featured-image                                     -2294 B  (-28.8%)       -373 B   (-8.7%)       -9 B   (-0.7%)
async-load-layout-masterbar-drafts                           -19213 B  (-16.6%)      -5741 B  (-10.2%)     -305 B   (-2.2%)
async-load-lib-abtest-test-helper                             -2252 B  (-34.5%)       -473 B  (-12.6%)      -43 B   (-3.9%)
async-load-lib-network-connection                              -848 B  (-18.3%)       -241 B   (-9.5%)      -49 B   (-5.0%)
async-load-lib-rubberband-scroll-disable                       -122 B  (-20.8%)        -20 B   (-3.9%)       -9 B   (-2.7%)
async-load-my-sites-current-site-domain-warnings              -6971 B  (-12.5%)      -3314 B   (-9.8%)    -1020 B  (-13.0%)
async-load-my-sites-guided-transfer                          -14118 B  (-24.5%)      -3459 B  (-10.7%)     -102 B   (-1.8%)
async-load-my-sites-site-settings-section-export             -16437 B  (-22.1%)      -4613 B  (-11.5%)     -255 B   (-3.3%)
async-load-my-sites-site-settings-section-import             -30671 B  (-17.5%)      -7470 B   (-7.9%)     -228 B   (-1.2%)
async-load-my-sites-stats-activity-log                       -55534 B  (-18.6%)     -14214 B   (-9.1%)     -557 B   (-1.9%)
async-load-my-sites-stats-comment-follows                    -22813 B  (-14.1%)      -4989 B   (-6.1%)     -129 B   (-0.7%)
async-load-my-sites-stats-overview                           -28148 B  (-17.6%)      -6567 B   (-7.6%)     -128 B   (-0.7%)
async-load-my-sites-stats-site                               -67793 B  (-17.6%)     -15290 B   (-7.4%)     -595 B   (-1.4%)
async-load-my-sites-stats-stats-insights                     -68693 B  (-16.3%)     -16151 B   (-7.5%)     -383 B   (-0.9%)
async-load-my-sites-stats-stats-post-detail                  -39760 B  (-16.2%)      -9292 B   (-7.4%)     -333 B   (-1.3%)
async-load-my-sites-stats-summary                            -52441 B  (-16.8%)     -11553 B   (-7.1%)     -438 B   (-1.3%)
async-load-notifications                                     -53308 B  (-18.1%)     -16726 B  (-10.1%)    -1334 B   (-3.7%)
async-load-post-editor-editor-author                          -7609 B  (-19.7%)      -1738 B   (-8.6%)      -34 B   (-0.7%)
async-load-post-editor-editor-discussion                      -2176 B  (-24.5%)       -358 B   (-7.3%)       +7 B   (+0.5%)
async-load-post-editor-editor-location                        -5847 B  (-28.4%)      -1296 B  (-11.2%)      -33 B   (-1.2%)
async-load-post-editor-editor-post-formats-accordion          -4089 B  (-23.9%)       -921 B  (-10.3%)      -39 B   (-1.7%)
async-load-post-editor-editor-seo-accordion                   -2479 B  (-25.3%)       -416 B   (-7.3%)      +42 B   (+2.4%)
async-load-post-editor-editor-sharing-accordion              -11885 B  (-25.1%)      -3045 B  (-11.3%)     -132 B   (-2.5%)
async-load-post-editor-media-modal                          -121764 B  (-17.5%)     -29810 B   (-8.2%)    -1184 B   (-1.7%)
async-load-reader-conversations-stream                        -2468 B  (-22.3%)       -531 B   (-8.2%)      -21 B   (-1.1%)
async-load-reader-feed-stream                                -15466 B  (-23.2%)      -3694 B   (-9.8%)      -77 B   (-1.1%)
async-load-reader-following-manage                           -30615 B   (-6.3%)      -7081 B   (-3.3%)     -307 B   (-0.8%)
async-load-reader-list-stream                                 -6009 B  (-19.2%)      -2579 B  (-15.8%)     -306 B   (-8.3%)
async-load-reader-search-stream                              -21319 B   (-4.8%)      -5222 B   (-2.7%)     -186 B   (-0.5%)
async-load-reader-sidebar                                    -19217 B  (-20.5%)      -6402 B  (-12.4%)     -519 B   (-5.2%)
async-load-reader-site-stream                                -16249 B  (-22.5%)      -3977 B   (-9.7%)     -110 B   (-1.4%)
async-load-reader-tag-stream-main                             -9481 B  (-24.7%)      -2125 B  (-10.0%)      -81 B   (-1.9%)
async-load-reader-team-main                                    -413 B  (-53.0%)        -84 B  (-17.4%)      -29 B   (-9.2%)
auth                                                           -505 B  (-44.9%)       -125 B  (-19.9%)      -37 B  (-11.0%)
build                                                       -644391 B  (-14.0%)    -223359 B  (-10.0%)   -21264 B   (-4.7%)
checklist                                                    -11128 B  (-24.4%)      -2856 B  (-10.5%)     -131 B   (-2.2%)
checkout                                                    -179524 B  (-13.4%)     -54477 B   (-7.6%)    -5286 B   (-3.7%)
comments                                                     -67563 B  (-16.8%)     -19342 B   (-9.3%)    -1346 B   (-3.2%)
concierge                                                    -31112 B  (-29.0%)      -8107 B  (-14.0%)    -1073 B  (-10.9%)
customize                                                    -16528 B  (-14.6%)      -5895 B  (-12.9%)     -724 B   (-6.2%)
devdocs                                                     -513736 B  (-13.4%)    -137895 B   (-5.9%)    -8018 B   (-1.9%)
domain-connect-authorize                                      -6584 B  (-23.5%)      -1688 B  (-10.3%)      -54 B   (-1.5%)
domains                                                     -235334 B  (-17.2%)     -69132 B   (-9.1%)    -5008 B   (-3.7%)
happychat                                                    -63617 B  (-17.4%)     -18297 B   (-9.4%)    -1157 B   (-2.9%)
hello-dolly                                                   -4411 B  (-32.2%)      -1249 B  (-14.9%)      -89 B   (-3.8%)
help                                                         -99175 B  (-20.3%)     -27609 B  (-10.1%)    -2284 B   (-4.3%)
jetpack-connect                                             -104640 B  (-13.8%)     -33796 B   (-8.5%)    -2953 B   (-3.5%)
jetpack-onboarding                                           -22009 B  (-27.8%)      -5608 B  (-13.2%)     -139 B   (-1.9%)
login                                                        -39387 B  (-19.9%)     -10694 B   (-9.5%)     -911 B   (-4.5%)
mailing-lists                                                 -2377 B  (-19.8%)       -655 B   (-9.1%)      -17 B   (-0.9%)
manifest                                                         +0 B                   +0 B                 +7 B   (+0.2%)
me                                                           -55415 B  (-18.9%)     -14727 B   (-9.0%)     -739 B   (-2.3%)
media                                                       -112104 B  (-19.2%)     -27923 B   (-9.1%)    -1031 B   (-1.8%)
notification-settings                                       -100718 B  (-19.4%)     -26018 B   (-8.9%)    -1023 B   (-1.9%)
paladin                                                       -1587 B  (-37.8%)       -453 B  (-18.2%)      -38 B   (-4.6%)
people                                                       -81654 B  (-19.8%)     -19708 B   (-8.7%)     -767 B   (-1.8%)
plans                                                       -104171 B  (-14.5%)     -36626 B   (-9.8%)    -3155 B   (-4.1%)
plugins                                                     -109245 B  (-16.6%)     -30673 B   (-8.6%)    -2701 B   (-3.8%)
post-editor                                                 -370882 B   (-8.0%)     -96567 B   (-5.1%)    -5026 B   (-1.1%)
posts-custom                                                 -93230 B  (-20.3%)     -26251 B  (-11.0%)    -1431 B   (-3.2%)
posts-pages                                                 -109265 B  (-18.7%)     -30576 B  (-10.0%)    -1639 B   (-2.8%)
preview                                                       -1826 B  (-22.7%)       -449 B   (-8.9%)      -67 B   (-3.6%)
purchases                                                   -198787 B  (-14.4%)     -60982 B   (-8.2%)    -4950 B   (-3.5%)
reader                                                      -145307 B  (-19.4%)     -40002 B  (-10.1%)    -2981 B   (-3.8%)
security                                                    -122868 B  (-17.3%)     -31544 B   (-7.8%)    -2143 B   (-2.7%)
sensei                                                        -2003 B  (-38.0%)       -482 B  (-15.7%)      -41 B   (-4.2%)
settings                                                    -130388 B  (-16.1%)     -41609 B  (-10.4%)    -3622 B   (-4.5%)
settings-discussion                                          -38807 B  (-20.9%)     -10224 B  (-10.4%)     -312 B   (-1.9%)
settings-security                                            -59437 B  (-23.2%)     -14590 B  (-10.5%)     -587 B   (-2.6%)
settings-traffic                                             -72948 B   (-9.0%)     -20565 B   (-6.0%)    -1255 B   (-1.7%)
settings-writing                                            -101178 B  (-15.9%)     -25093 B   (-8.0%)    -1223 B   (-2.1%)
sharing                                                      -61616 B  (-17.9%)     -16537 B   (-8.8%)     -853 B   (-2.6%)
signup                                                      -155024 B  (-14.4%)     -48149 B   (-8.2%)    -4456 B   (-3.5%)
stats                                                         -5014 B  (-16.8%)      -2077 B  (-12.2%)     -122 B   (-3.1%)
theme                                                        -52386 B  (-18.4%)     -15153 B  (-10.9%)     -970 B   (-3.5%)
themes                                                       -82263 B  (-19.3%)     -22934 B  (-10.7%)    -1787 B   (-4.2%)
woocommerce                                                 -450710 B   (-9.6%)    -157137 B   (-7.6%)   -12834 B   (-3.0%)
wp-job-manager                                               -50584 B  (-22.3%)     -16423 B  (-13.3%)    -1125 B   (-5.3%)
wp-super-cache                                               -63661 B  (-19.2%)     -19840 B  (-11.2%)    -1060 B   (-3.4%)
zoninator                                                    -41054 B  (-22.5%)     -13471 B  (-13.8%)    -1091 B   (-6.6%)
account                                                      -77597 B  (-17.7%)     -19901 B   (-8.2%)     -639 B   (-1.4%)
```

gziolo commented:
21264 B for the `build` file. Nice :)

samouri commented:
I've updated this PR to only encompass esmodule activation and tree shaking. We can explore scope-hoisting easily enough in a followup pr.

dmsnell commented:
http://iscalypsofastyet.com/branch?branch=try/scope-hoisting

```
Delta:
chunk                                                       stat_size            parsed_size            gzip_size
accept-invite                                                -35435 B  (-17.2%)      -8647 B   (-7.5%)     -566 B   (-2.4%)
account-recovery                                             -22300 B  (-27.1%)      -6943 B  (-15.1%)    -1096 B  (-12.4%)
ads                                                          -26712 B  (-19.2%)      -7775 B   (-9.7%)     -258 B   (-1.7%)
async-load-blocks-calendar-popover                           -11372 B   (-3.3%)      -2712 B   (-1.6%)      -57 B   (-0.2%)
async-load-blocks-reader-full-post                           -22594 B  (-20.1%)      -6338 B  (-10.3%)     -434 B   (-3.2%)
async-load-components-happychat                               -9270 B  (-12.2%)      -3872 B  (-11.2%)     -382 B   (-4.3%)
async-load-components-web-preview-component                  -98500 B  (-18.9%)     -25535 B   (-9.4%)    -1567 B   (-2.8%)
async-load-components-webpack-build-monitor                    -966 B  (-17.1%)       -117 B   (-4.0%)       -3 B   (-0.2%)
async-load-extensions-woocommerce-app-store-stats            -25994 B   (-8.0%)     -31221 B  (-14.9%)    -7149 B  (-14.9%)
async-load-extensions-woocommerce-app-store-stats-listview   -11290 B  (-25.1%)      -2523 B  (-10.2%)      -72 B   (-1.4%)
async-load-featured-image                                     -2294 B  (-28.8%)       -373 B   (-8.7%)       -9 B   (-0.7%)
async-load-layout-masterbar-drafts                           -19246 B  (-16.6%)      -5741 B  (-10.2%)     -306 B   (-2.3%)
async-load-lib-abtest-test-helper                             -2252 B  (-34.5%)       -473 B  (-12.6%)      -43 B   (-3.9%)
async-load-lib-network-connection                              -846 B  (-18.5%)       -241 B   (-9.6%)      -49 B   (-5.1%)
async-load-lib-rubberband-scroll-disable                       -122 B  (-20.8%)        -20 B   (-3.9%)       -9 B   (-2.7%)
async-load-my-sites-current-site-domain-warnings              -6971 B  (-12.5%)      -3314 B   (-9.8%)    -1020 B  (-13.0%)
async-load-my-sites-guided-transfer                          -14277 B  (-24.9%)      -3465 B  (-10.7%)     -114 B   (-2.0%)
async-load-my-sites-site-settings-section-export             -16698 B  (-22.5%)      -4625 B  (-11.6%)     -268 B   (-3.4%)
async-load-my-sites-site-settings-section-import             -30989 B  (-17.7%)      -7482 B   (-7.9%)     -233 B   (-1.2%)
async-load-my-sites-stats-activity-log                       -55477 B  (-18.6%)     -14110 B   (-9.1%)     -543 B   (-1.9%)
async-load-my-sites-stats-comment-follows                    -22799 B  (-14.1%)      -4995 B   (-6.1%)     -132 B   (-0.7%)
async-load-my-sites-stats-overview                           -28134 B  (-17.6%)      -6573 B   (-7.7%)     -134 B   (-0.8%)
async-load-my-sites-stats-site                               -68042 B  (-17.6%)     -15345 B   (-7.4%)     -610 B   (-1.4%)
async-load-my-sites-stats-stats-insights                     -69085 B  (-16.4%)     -16147 B   (-7.5%)     -379 B   (-0.9%)
async-load-my-sites-stats-stats-post-detail                  -39787 B  (-16.2%)      -9304 B   (-7.4%)     -325 B   (-1.3%)
async-load-my-sites-stats-summary                            -52427 B  (-16.8%)     -11559 B   (-7.1%)     -436 B   (-1.3%)
async-load-notifications                                     -53308 B  (-18.1%)     -16726 B  (-10.1%)    -1334 B   (-3.7%)
async-load-post-editor-editor-author                          -7609 B  (-19.7%)      -1738 B   (-8.6%)      -34 B   (-0.7%)
async-load-post-editor-editor-discussion                      -2176 B  (-24.5%)       -358 B   (-7.3%)       +7 B   (+0.5%)
async-load-post-editor-editor-location                        -5847 B  (-28.4%)      -1296 B  (-11.2%)      -33 B   (-1.2%)
async-load-post-editor-editor-post-formats-accordion          -4125 B  (-24.3%)       -921 B  (-10.3%)      -43 B   (-1.9%)
async-load-post-editor-editor-seo-accordion                   -2479 B  (-25.3%)       -416 B   (-7.3%)      +42 B   (+2.4%)
async-load-post-editor-editor-sharing-accordion              -11885 B  (-25.1%)      -3045 B  (-11.3%)     -132 B   (-2.5%)
async-load-post-editor-media-modal                          -122444 B  (-17.6%)     -29840 B   (-8.3%)    -1183 B   (-1.7%)
async-load-reader-conversations-stream                        -2468 B  (-22.3%)       -537 B   (-8.3%)      -23 B   (-1.2%)
async-load-reader-feed-stream                                -15480 B  (-22.8%)      -3722 B   (-9.7%)      -76 B   (-1.1%)
async-load-reader-following-manage                           -30619 B   (-6.2%)      -7103 B   (-3.3%)     -313 B   (-0.8%)
async-load-reader-list-stream                                 -6256 B  (-20.1%)      -2591 B  (-15.9%)     -305 B   (-8.3%)
async-load-reader-search-stream                              -21323 B   (-4.8%)      -5244 B   (-2.7%)     -185 B   (-0.5%)
async-load-reader-sidebar                                    -19481 B  (-20.9%)      -6402 B  (-12.5%)     -513 B   (-5.1%)
async-load-reader-site-stream                                -16263 B  (-22.2%)      -4005 B   (-9.6%)     -117 B   (-1.5%)
async-load-reader-tag-stream-main                             -9538 B  (-24.9%)      -2131 B  (-10.1%)      -81 B   (-1.9%)
async-load-reader-team-main                                    -413 B  (-53.0%)        -84 B  (-17.4%)      -29 B   (-9.2%)
auth                                                           -505 B  (-44.9%)       -167 B  (-26.6%)      -49 B  (-14.6%)
build                                                       -623337 B  (-13.9%)    -251614 B  (-11.7%)   -26413 B   (-6.0%)
checklist                                                    -11128 B  (-24.4%)      -2890 B  (-10.6%)     -142 B   (-2.4%)
checkout                                                    -180940 B  (-13.5%)     -55222 B   (-7.7%)    -5420 B   (-3.8%)
comments                                                     -68062 B  (-16.9%)     -19440 B   (-9.4%)    -1363 B   (-3.3%)
concierge                                                    -31112 B  (-29.0%)      -8175 B  (-14.1%)    -1092 B  (-11.1%)
customize                                                    -16811 B  (-14.9%)      -5929 B  (-13.0%)     -737 B   (-6.3%)
devdocs                                                     -516116 B  (-13.5%)    -138159 B   (-5.9%)    -7997 B   (-1.9%)
domain-connect-authorize                                      -6584 B  (-23.5%)      -1724 B  (-10.5%)      -65 B   (-1.8%)
domains                                                     -236084 B  (-17.3%)     -70318 B   (-9.3%)    -5132 B   (-3.8%)
happychat                                                    -64423 B  (-17.7%)     -18662 B   (-9.6%)    -1202 B   (-3.0%)
hello-dolly                                                   -4411 B  (-32.2%)      -1269 B  (-15.1%)      -99 B   (-4.2%)
help                                                        -100260 B  (-20.6%)     -28000 B  (-10.3%)    -2353 B   (-4.5%)
jetpack-connect                                             -107205 B  (-14.0%)     -35156 B   (-8.8%)    -3225 B   (-3.8%)
jetpack-onboarding                                           -23199 B  (-27.5%)      -5869 B  (-12.9%)     -130 B   (-1.7%)
login                                                        -40322 B  (-20.4%)     -10805 B   (-9.6%)     -957 B   (-4.7%)
mailing-lists                                                 -2377 B  (-19.8%)       -669 B   (-9.3%)      -22 B   (-1.1%)
manifest                                                         +0 B                   +0 B                 +1 B   (+0.0%)
me                                                           -56221 B  (-19.3%)     -15176 B   (-9.3%)     -797 B   (-2.5%)
media                                                       -112934 B  (-19.4%)     -27963 B   (-9.1%)    -1038 B   (-1.9%)
notification-settings                                       -101524 B  (-19.6%)     -26455 B   (-9.1%)    -1109 B   (-2.1%)
paladin                                                       -1587 B  (-37.8%)       -481 B  (-19.3%)      -45 B   (-5.4%)
people                                                       -83180 B  (-19.8%)     -19890 B   (-8.7%)     -776 B   (-1.8%)
plans                                                       -104669 B  (-14.6%)     -36890 B   (-9.9%)    -3198 B   (-4.2%)
plugins                                                     -109426 B  (-16.7%)     -30939 B   (-8.7%)    -2735 B   (-3.9%)
post-editor                                                 -372489 B   (-8.0%)     -96632 B   (-5.1%)    -5045 B   (-1.1%)
posts-custom                                                 -94166 B  (-20.5%)     -26293 B  (-11.0%)    -1448 B   (-3.3%)
posts-pages                                                 -110201 B  (-18.9%)     -30628 B  (-10.0%)    -1657 B   (-2.9%)
preview                                                       -1826 B  (-22.7%)       -470 B   (-9.4%)      -76 B   (-4.1%)
purchases                                                   -200260 B  (-14.6%)     -61681 B   (-8.3%)    -5119 B   (-3.6%)
reader                                                      -145406 B  (-19.4%)     -40228 B  (-10.1%)    -2996 B   (-3.8%)
security                                                    -124078 B  (-17.5%)     -32329 B   (-8.0%)    -2222 B   (-2.8%)
sensei                                                        -2003 B  (-38.0%)       -502 B  (-16.3%)      -51 B   (-5.2%)
settings                                                    -131366 B  (-16.2%)     -41813 B  (-10.5%)    -3647 B   (-4.5%)
settings-discussion                                          -39265 B  (-21.2%)     -10250 B  (-10.4%)     -312 B   (-1.9%)
settings-security                                            -59952 B  (-23.5%)     -14538 B  (-10.5%)     -577 B   (-2.5%)
settings-traffic                                             -74023 B   (-9.1%)     -21594 B   (-6.3%)    -1287 B   (-1.7%)
settings-writing                                            -101803 B  (-16.1%)     -25133 B   (-8.0%)    -1208 B   (-2.1%)
sharing                                                      -62435 B  (-18.2%)     -16585 B   (-8.8%)     -874 B   (-2.7%)
signup                                                      -156105 B  (-14.5%)     -48073 B   (-8.2%)    -4475 B   (-3.5%)
stats                                                         -5065 B  (-17.0%)      -2477 B  (-14.6%)     -167 B   (-4.3%)
theme                                                        -52830 B  (-18.6%)     -15180 B  (-10.9%)     -985 B   (-3.6%)
themes                                                       -82697 B  (-19.5%)     -23042 B  (-10.7%)    -1802 B   (-4.2%)
woocommerce                                                 -451640 B   (-9.7%)    -157359 B   (-7.6%)   -12919 B   (-3.0%)
wp-job-manager                                               -50935 B  (-22.5%)     -16498 B  (-13.4%)    -1135 B   (-5.4%)
wp-super-cache                                               -64089 B  (-19.3%)     -19874 B  (-11.2%)    -1074 B   (-3.4%)
zoninator                                                    -41308 B  (-22.7%)     -13531 B  (-13.8%)    -1111 B   (-6.7%)
account                                                      -78413 B  (-17.9%)     -20378 B   (-8.4%)     -701 B   (-1.5%)
```

samouri commented:
Summary of changes by commit:

**Build: disable commonjs transform**:

While the overall goal for this commit is to disable the commonjs module transformation, I couldn't do it by following the most straightforward path.  The babel [docs](https://babeljs.io/docs/plugins/preset-env/#modules) show that you can disable the transform by setting `modules: false` as an option to the "env" preset.

Initially I tried simply editing the `.babelrc`, but that blew up because we have a number of scripts as well as our jest tests that share the same babel config and need the transformation. Therefore I edited just the config that gets used by webpack.



Two other changes that needed to happen as well:

1. Removed the transform `add-module-exports`.  The transform was intended to allow us to have an easier time importing es6 modules using `require`, even though the code we were writing doesn't match spec.  It allowed us to use a default export like this: `const SomeComponent = require( 'some-component' )` whereas technically we should have been writing: `const SomeComponent = require( 'some-component ).default`.  I wanted to remove it from .babelrc but almost all of our jest tests rely upon this functionality, so instead I manually remove it from both webpack configs. Fixing the fallout of this removal is what the next commit is all about.  
2. Moved `inline-imports` to only exist in webpack.config.js instead of babelrc. I don't know why, but changing the build settings also affected how the module is resolved. jest expects no file extensions whereas webpack expected it -- and in any case, the transform isn't relevant for jest or the server.

**require --> import + require.default**

This commit fixes almost all of the issue caused by the previous commit.  For the most part it changes any latent require statements to imports, and then adds `.default` to the require statements that cannot be easily updated. I left two special cases for the next two commits

**sections-middleware**: sections-middleware.js was pretty simple to fix. Since sections-middleware calls `load()`  which is based on [dynamic import](http://2ality.com/2017/01/import-operator.html#accessing-default-exports), it needs to explicitly call the default member.

**Extensions**: The extensions-loader is used by both the server and client.  Unlike the sections-loader though, it doesn't switch between import and require depending on which it is in.  In order to compensate, I manually map all of the reducers to .default on  boot.

**Fix test**: after all these commits a single test was failing.

bsessions85 commented:
e2e tests passed

samouri commented:
I've created a `wp-desktop` fix, but its pretty gross. 
It may be worthwhile to look into whether or not we can selectively include `add-module-exports` for jest and then remove it from babelrc

jsnajdr commented on client/boot/common.js:
```diff
@@ -22,6 +22,9 @@ import { setLocale, setLocaleRawData } from 'state/ui/language/actions';
 import { getSections, setupRoutes } from 'sections-middleware';
+import { checkFormHandler } from 'lib/protect-form';
+import notices from 'notices';
+import authController from 'auth/controller';
```
Hah, the `notices` and `authController` modules still use `export default = { ... }` :)

samouri replied:
just double checking: this comment is just an observation and doesn't imply that i should make a change here right?

jsnajdr replied:
Yes, it can wait for a separate PR.

jsnajdr commented on client/lib/happychat/connection-ng.js:
```diff
@@ -23,7 +24,7 @@ import {
 } from 'state/happychat/connection/actions';
 
-const debug = require( 'debug' )( 'calypso:happychat:connection' );
+const debug = debugFactory()( 'calypso:happychat:connection' );
```
This should be `debugFactory( 'calypso:happychat:connection' )`. Otherwise you create a debug logger and log a message immediately :)

samouri replied:
whoops! Nice catch. I'll  fixem

jsnajdr commented on client/lib/happychat/connection.js:
```diff
@@ -23,7 +24,7 @@ import {
 } from 'state/happychat/connection/actions';
 
-const debug = require( 'debug' )( 'calypso:happychat:connection' );
+const debug = debugFactory()( 'calypso:happychat:connection' );
```
same here.

jsnajdr commented:
There's a `require( 'wpcom-proxy-request' )` in `client/lib/wpcom-undocumented/lib/me.js` that could also be updated to `import`.

dmsnell commented:
@jsnajdr _et. al._ @samouri and I had a conversation in Slack concerning the change in semantics from runtime `require()` and module-init-time `import` statements. In most cases it won't matter (and I don't know right now if it matters for `wpcom-proxy-request` but my guess is that it does) but there are reasons to leave it as `require()` until we have `import()` running.

Namely the `import` statements load and initialize their referenced modules as soon as the calling module initializes. While the code may end up in the bundle either way, there are places where we can't execute that code because of conditions which aren't met in some runtime circumstances. In something like `wpcom-proxy-request` it may not matter but if we don't need it then we have no reason to initialize that module anyway.

jsnajdr commented:
> I don't know right now if it matters for `wpcom-proxy-request` but my guess is that it does

The `wpcom-proxy-request` module does very little static initialization: it sets a few variables to `false` or `null`, computes `origin` from `window.location` and sends a probe request to `window.postMessage` to determine if it supports non-string messages. All the heavy lifting, i.e., initializing the proxy iframe, is done lazily on first call to `request`.

But you have a very good point: for the modules that this PR changes from conditional imports to static, we should check if any of them does some expensive static initialization.

If it does, I think the most reliable remedy is to rewrite these modules to do the expensive initialization lazily -- only when an exported function is called or exported class instantiated for the first time.

jsnajdr commented:
I found an insidious bug that's exposed by enabling tree shaking and scope hoisting: in `lib/posts/utils.js`, there is `isPublished` function that internally calls `this.isBackDatedPublished`, but there is no `this` binding.

It used to work when the `isPublished` and `isBackDatedPublished` functions were properties of one `exports` object, but now they are defined as standalone functions in the global scope.

To experience the runtime error, go to the post editor page at `/post/:site_id/:post_id`.

How many other modules are affected and where to find them?

samouri commented:
> It used to work when the isPublished and isBackDatedPublished functions were properties of one exports object, but now they are defined as standalone functions in the global scope.

So the solution should be to remove the `this` right?

jsnajdr commented:
> So the solution should be to remove the `this` right?

Yes.

samouri commented:
> How many other modules are affected and where to find them?

Specifically for the "this" bug you've found, we can do a grep for "this" for all `*.js` (non jsx) files.
This should at least give us a sense for how often this practice is done in Calypso.

The result of the query was 17 files. They were grouped like this:

- `bin/*.js`: 9 scripts within `bin/`.  They each reference "this" within comments. These are safe.
- `client/lib/happychat`: 2 cases. Both of them are because these files actually use classes. This is pretty unusual since most of our classes live in jsx files. These two should also be safe.
- `client/lib/post`: 1 case. This is the bug you found.
- `sections-preload.js`, `wordpress-com.js`, `inline-imports.js`: uses the word this in a comment. safe.
- webpack configs: both use the word this a couple of times, also safe.

It seems to me like you caught something pretty rare. Technically its possible this also happens within jsx files, but I wouldn't know how to search for that.  @dmsnell knows of some structural grep tools that could help.

----

It is worrisome to think that there still may be other slight inconsistencies between how CJS and esmodules behave in webpack.  I don't think it should stop us progressing with this PR though.

dmsnell commented:
> sends a probe request to window.postMessage

that's the kind of thing we need to be careful about because if we initialize that via `import` when it was previously trapped behind a `window` check then we'll throw an exception out of the browser environment (server-renderings etc…).

> Insidious bug…It used to work when the isPublished and isBackDatedPublished functions were properties of one exports object, but now they are defined as standalone functions in the global scope.

Good catch! This isn't inherently a problem though; it was just missed in the conversion (probably by me). We rewrote many functions to go from `this.something` or `Something.somethingElse` to non-prefixed `something()`s. We just need to update the code.

From what I've seen we have very few cases (outside of React components) where we need `this` or rely on its behavior (that is, where we couldn't just as well rewrite it without `this`).

> structural grep tools that could help

maybe? these might be hard to find in the JSX files. on the other hand, if it's in a JSX file it's probably safe because `this` wouldn't extend outside of the library. my guess from a practical standpoint is that if there are any more of these out there they will surface and we will fix them.

> It is worrisome to think that there still may be other slight inconsistencies

very much so. this is also why I think in this case code-modding would have been disastrous in practice.

> I don't think it should stop us progressing with this PR though.

:+1:

samouri commented:
@timmyc: I know I'm referencing code from about 2 years ago, but I was wondering if you remember what the intention of `getPagePath()` was? I make a modification in the latest commit to make it pass the tests and want to make sure I'm not changing behavior in a way that would cause trouble: https://github.com/Automattic/wp-calypso/pull/16057/commits/c2404e9403d022ac1354dcee65ebbdd060208b54

samouri commented:
Actually, its only used in a single line in the whole app: `client/post-editor/post-editor.jsx`.

timmyc commented:
Memory lane right there, well it was originally used for displaying the correct base path for post slugs. Your changes seem fine to me.

jsnajdr commented on client/lib/posts/utils.js:
```diff
@@ -179,33 +190,23 @@ export const getPermalinkBasePath = function( post ) {
-
-	return pathParts.join( '/' );
+	const postUrl = get( post, [ 'other_URLs', 'permalink_URL' ] ) || post.URL;
+	return removeSlug( postUrl );
```
... and you can also revert this back to just `return removeSlug( post.URL )` -- the permalink case was already handled.

samouri replied:
While I think thats true, it'll fail one of the unit tests.

samouri replied:
edit: i lied. it works fine now that the `return` was added back :)
